compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -2818,6 +2803,56 @@ private BType checkIndexAccessExpr(BLangIndexBasedAccess indexBasedAccessExpr, B
         return actualType;
     }
 
+    private BType checkTupleIndexBasedAccess(BLangIndexBasedAccess accessExpr, BType indexExprType, BTupleType tuple) {
+        BType actualType = symTable.semanticError;
+        BLangExpression indexExpr = accessExpr.indexExpr;
+        switch (indexExprType.tag) {
+            case TypeTags.INT:
+                if (indexExpr.getKind() == NodeKind.NUMERIC_LITERAL) {
+                    int indexValue = ((Long) ((BLangLiteral) indexExpr).value).intValue();
+                    actualType = checkTupleFieldType(tuple, indexValue);
+                    if (actualType.tag == TypeTags.SEMANTIC_ERROR) {
+                        dlog.error(accessExpr.pos,
+                                DiagnosticCode.TUPLE_INDEX_OUT_OF_RANGE, indexValue, tuple.tupleTypes.size());
+                    }
+                } else {
+                    BTupleType tupleExpr = (BTupleType) accessExpr.expr.type;
+                    LinkedHashSet<BType> tupleTypes = new LinkedHashSet<>(tupleExpr.tupleTypes);
+                    actualType = tupleTypes.size() == 1 ? tupleTypes.iterator().next() :
+                            new BUnionType(null, tupleTypes, false);
+                }
+                break;
+            case TypeTags.FINITE:
+                BFiniteType finiteIndexExpr = (BFiniteType) indexExprType;
+                LinkedHashSet<BType> possibleTypes = new LinkedHashSet<>();
+                for (BLangExpression finiteMember : finiteIndexExpr.valueSpace) {
+                    if (finiteMember.type.tag != TypeTags.INT) {
+                        dlog.error(indexExpr.pos, DiagnosticCode.INVALID_FINITE_TYPE_INDEX_EXPR, indexExprType);
```

Comment:
Something like this will fail, right - 

```ballerina
type V 0|"x";
V v = 0;

(int, string) t = (1, "one");
int i = t[v];
```

Is that the correct behavior according to the spec? This looks like a valid scenario.

Reply:
Index expressions are not completed yet in the Spec. Went with this restrictive approach so as to avoid adding more checks at the runtime. Added a test case for it as well. Yes, it is valid but personally didn't see a critical use case for such a scenario.

Reply:
Yup, it is not critical. Maybe later we can allow it.